pkg/security/podsecuritypolicy/apparmor/withapparmor.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apparmor
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/security/apparmor"
+	"k8s.io/kubernetes/pkg/util/validation/field"
+)
+
+type withAppArmorProfile struct {
+	allowedProfiles []string
+}
+
+var _ AppArmorStrategy = &withAppArmorProfile{}
+
+// NewWithAppArmorProfile creates a new strategy that enforces AppArmor profile constraints.
+func NewWithAppArmorProfile(allowedProfiles []string) AppArmorStrategy {
+	return &withAppArmorProfile{allowedProfiles}
+}
+
+func (s *withAppArmorProfile) Generate(pod *api.Pod, container *api.Container) error {
+	if apparmor.GetProfileName(pod, container.Name) != "" {
+		// Profile already set, nothing to do.
+		return nil
+	}
+
+	if len(s.allowedProfiles) == 0 {
+		// AppArmor disabled, don't set a profile.
+		return nil
+	}
+
+	apparmor.SetProfileName(pod, container.Name, s.allowedProfiles[0])
```

Comment:
Generally I've tried to keep these methods just returning the value of what is generated and not actually setting anything.  Then the provider applies the values to the copy of the SecurityContext, or in this case annotations.  I'm also leaning in favor of changing the provider interface to return a set of annotations that I mentioned over email since it lets us keep the assumption that the provider does not actually change the pod and it is up to the caller to decide if it should apply the generated values.